### PR TITLE
[RLlib] Fix memory leak in APEX_DQN

### LINKED ISSUE #26580: [RLlib Release Tests] APEX long running release test is failing
### What happened + What you expected to happen

n/a

### Versions / Dependencies

master

### Reproduction script

n/a

### Issue Severity

_No response_

### PR DESCRIPTION
**Pull request by avnishn**
There was some sort of memory consumption issue in the queue
that we use for reading replay batches in to before placing them
on the learner queue. To be honest I can't exactly articulate the
bug, but its definitely some issue with this list object.

I also made the queue placement operation blocking, which is similar
to the other PR I opened this week with regards to making queue
placement operations blocking #26581.

Signed-off-by: avnish <avnish@anyscale.com>

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by avnishn** [comment]
I kicked off a set of release tests for the tests that might be affected (learning a-e and long running apex), lets see what happens...

**Comment by avnishn** [comment]
TLDR; it could be an OOM from some misuse of a list containing ray objects, or it could just be that this list that we use to store sample batches from replay buffers before placing them on the learner queue grows unboundedly, which would be a problem anyways

**Comment by ArturNiederfahrenhorst** [comment]
A-E are marked green but are failing.

**Comment by avnishn** [comment]
The failure is unrelated (a flaky test that needs to be fixed) The long running apex is still running, but has been running for 15 hours, so I think its safe to assume that it is fine. It was terminating in 30 minutes before this (when it would fail) and can run for 24 hours iirc.

File: rllib/tuned_examples/apex_dqn/pong-apex-dqn.yaml [hunk 1]
```diff
@@ -18,5 +18,8 @@ pong-apex:
         replay_buffer_config:
           type: MultiAgentPrioritizedReplayBuffer
           capacity: 1000000
+        # we should set this because few machines would be able to contain
+        # the replay buffers in memory otherwise
+        compress_observations: True
```
Comment (gjoliver) [inline rllib/tuned_examples/apex_dqn/pong-apex-dqn.yaml:23]: doesn't this change a lot of things?
we will be spending tons of CPU cycles on compression and decompression right?

File: rllib/algorithms/apex_dqn/apex_dqn.py [hunk 2]
```diff
@@ -583,21 +582,12 @@ def wait_on_replay_actors() -> None:
...
-                # the minimum threshold yet.
-                if item:
-                    self.learner_thread.inqueue.put(
-                        self.replay_sample_batches[0], timeout=0.001
-                    )
-                    self.replay_sample_batches.pop(0)
-            except queue.Full:
-                break
+        for item in replay_sample_batches:
+            self.learner_thread.inqueue.put(item, block=True)
```
Comment (gjoliver) [inline rllib/algorithms/apex_dqn/apex_dqn.py:589]: I feel like this is essentially back-pressure. like if learner queue if full, and learner has trouble keeping up with the input sample, we stop sampling from RB.
can we print the size of self.replay_sample_batches as a metric in the old world without this fix?
that way, we can easily verify that this list growing un-boundedly was the reason for mem-leak-ish behavior.

**Review by gjoliver — CHANGES_REQUESTED** [review]
there is a push for always having tests when submitting fixes, but I guess it's gonna be really hard to write a unit test for this PR huh ...

**Comment by avnishn** [comment]
following up in @gjoliver's request:

> I feel like this is essentially back-pressure. like if learner queue if full, and learner has trouble keeping up with the input sample, we stop sampling from RB.
can we print the size of self.replay_sample_batches as a metric in the old world without this fix?
that way, we can easily verify that this list growing un-boundedly was the reason for mem-leak-ish behavior.

image


here's the size of self.replay_sample_batches over time. The memory bug is directly attributed to this data structure growing in an unbounded fashion

**Comment by avnishn** [comment]
following up I removed the change to the yaml file for a separate pr. 

I don't think we need to write a unit test. The whole point of this pr to begin with was to fix the not passing test, which is checking for the exact fix that we're providing here. @gjoliver

**Comment by avnishn** [comment]
@gjoliver see #26735

File: rllib/tuned_examples/apex_dqn/pong-apex-dqn.yaml (hunk 1 above)
Comment (avnishn) [reply rllib/tuned_examples/apex_dqn/pong-apex-dqn.yaml]: removed this and opened a separate pr

**Comment by gjoliver** [comment]
awesome man!! thanks for figuring this one out.
yeah, I think these multi threading things are kinda hard to unit test.

**Review by gjoliver — APPROVED** [review]
